presto-main/src/main/java/com/facebook/presto/operator/repartition/BlockEncodingBuffer.java

```diff
@@ -50,8 +50,9 @@
 
     /**
      * Signals that this is the last batch in this page, so that the internal buffers in BlockEncodingBuffers can recycled.
+     * @param bufferFull
      */
-    void noMoreBatches();
+    void noMoreBatches(boolean bufferFull);
```

Comment:
Do we really need this new parameter here? resetBuffer() seems to be signaling the same condition. Can't values buffer be release in resetBuffers()?

Reply:
@mbasmanova There are two conditions that have to be met to recycle these buffers: 1) bufferFull == true (just flushed);  2) This is the last batch in this page. If condition 2) is not met, the buffers are still needed for the next batch in the same page. Even if we returned these buffers if only 1) is met, the next batch would borrow these buffers again. Borrowing/Returning the buffers are not free because of the Deque and Set operations are expensive. Therefore we only want to recycle the buffers when both of these two conditions are met. resetBuffer() only happens when flushing, but it doesn't guarantee that it's the last batch. Another way to do this is to signal the `lastBatchInPage` in `flush(lastBatchInPage)` but since we have done it in the `finally` section for the int[] buffers, I think it's better to do it the same way at the same place for the byte[] buffers.

Reply:
@yingsu00 Thanks for explaining. Sounds like the object itself can figure out whether bufferFull is true or not, e.g. it can set a internal flag in resetBuffer, use it in noMoreBatches and reset it in setupDecodedBlocksAndPositions.

Reply:
@mbasmanova I see. This requires adding an internal flag in AbstractBlockEncodingBuffer to signal if the buffers are just flushed or not. I'll make the change as you suggested.

Reply:
@mbasmanova I just updated the PR using the way you suggested. One thing to mention is that I realized that we can't reset the `flushed` flag in setupDecodedBlocksAndPositions(), but instead, it has to be reset at the beginning of the `do...while` loop e.g. `setNextBatch`. Every batch needs to begin with the not-flushed state. Other than this, other changes were the same as you suggested. Thanks for reviewing and bringing up the valuable suggestion!